### DISCUSSION TITLE
Can I run semantic-release to just jump the version?

### DISCUSSION THREAD

**Discussion by javierguzman** [semantic-release/semantic-release#2987] [Q&A]
Hello all,

I have some Github actions setup where I was running the following jobs:
1. testing
2. build and generate executables
3. run semantic-release to generate release

I have realised that I was building the executables with the old version, as semantic-release was running in the last part. So I have modified the CI/CD like this:
1. testing
2. run semantic-release to generate release
3. build and generate executables
4. run semantic-release to generate release

However, this fails because it tries to upload the release but it does not exist. Is there a way to achieve what I need? Meaning, bump the version, then build, then create and upload the release.

This is my release.config.js:

```
module.exports = {
  branches: ["main"],
  plugins: [
    "@semantic-release/commit-analyzer",
    "@semantic-release/release-notes-generator",
    "@semantic-release/changelog",
    "@semantic-release/npm",
    "@semantic-release/git",
    [
      "@semantic-release/github",
      {
        prepare: [
          {
            path: "@semantic-release/exec",
            cmd: "mv dist/*.exe dist/myapp-v${nextRelease.version}.exe",
          },
          {
            path: "@semantic-release/exec",
            cmd: "mv dist/*-x64.dmg dist/myapp-v${nextRelease.version}-x64.dmg",
          },
          {
            path: "@semantic-release/exec",
            cmd: "mv dist/*-arm64.dmg dist/myapp-v${nextRelease.version}-arm64.dmg",
          },
          {
            path: "@semantic-release/exec",
            cmd: "mv dist/*.AppImage dist/myapp-v${nextRelease.version}.AppImage",
          },
        ],
        assets: [
          {
            path: "dist/myapp-v${nextRelease.version}.exe",
            name: "myapp-v${nextRelease.version}.exe",
          },
          {
            path: "dist/myapp-v${nextRelease.version}-x64.dmg",
            name: "myapp-v${nextRelease.version}-x64.dmg",
          },
          {
            path: "dist/myapp-v${nextRelease.version}-arm64.dmg",
            name: "myapp-v${nextRelease.version}-arm64.dmg",
          },
          {
            path: "dist/myapp-v${nextRelease.version}.AppImage",
            name: "myapp-v${nextRelease.version}.AppImage",
          },
        ],
      },
    ],
  ],
};
```

Thank you in advance and regards

**Comment by travi** [comment]
semantic-release is intended to manage your entire release workflow, which should ideally include any build steps. with an npm project, the simplest way to accomplish this is with a `prepack` script so that your build automatically runs before the publish step

**Reply by javierguzman** [reply]
Thanks @travi for your quick response! Isn't there any other way or alternative? I am just asking because I have already the CI/CD built with Github Actions and more importantly I am building the app (Desktop app) for MacOS, Linux and Windows. So I can think of two main problems:
1. I am not sure I can insert all that CI/CD logic within the semantic-release script, because for example I even need to run java to sign the executables
2. I need to run the script for Linux, Windows and MacOS meaning I would bump the version 3 times if I am not mistaken or as minimum insert complex logic to get around that

Again thank you

**Reply by travi** [reply]
> I need to run the script for Linux, Windows and MacOS meaning I would bump the version 3 times if I am not mistaken or as minimum insert complex logic to get around that

this is a detail that does have complications that we do not have a good solution for in semantic-release today. my statement still stands about the goal of semantic-release being to handle the entire release process. this basically gives two options:
* let semantic-release handle everything, including build steps. this is our recommendation any time that it is possible
* do your build steps and then let semantic-release make any updates to the built files that are necessary, using plugins like `@semantic-release/exec` as @seebeen mentioned below, or https://github.com/gr2m/semantic-release-plugin-update-version-in-files if specifically looking to update versions. this can be trickier than just building when semantic-release can provide a version, so we normally advise against it.

since semantic-release can only run from a single workflow run, you are looking at the second option

**Reply by seebeen** [reply]
@travi 'member when I told you a flag that would just output the next version number to support multi-platform releases? 😅 - This is the perfect use case. Having something that can just determine the next version string and enable users to implement it further solves almost all of multi-stage / multi-platform builds and releases.

#1647 Constantly coming back to us 😅

@javierguzman The fastest way to solve your problem of determining the next version would be this. 

When you need the next release version for the builds you can do this:
```bash
NEXT_VERSION=$(npx run semantic-release --dry-run | grep -oP 'Published release \K.*? ')
```
Read the docs about [Dry Run](https://semantic-release.gitbook.io/semantic-release/usage/configuration#dryrun)

This will run the semantic release in a dry run mode, and will give you the next release version you can use in the following steps.

Once you have your builds - you can run semantic-release proper.

Hope this helps.

**Reply by javierguzman** [reply]
Thanks @travi and @seebeen and sorry for the delay, I meant to answer on Friday but I was suddenly required when I was writing this message. I have checked your suggestion and if I am not mistaken that would not commit anything whatsoever, so I think the workflow would be this one:
1. Grab next version
2. Somehow update the package.json version based on the next version
3. Build everything 
4. Make release with semantic-release without bumping the version

So my question for this workflow would be, is it possible to run semantic-release without bumping the version? So I can call it for point 4?

Does this make sense?

Again thank you for all the help

**Comment by seebeen** [comment]
Hello. I think I can help you out. I'm doing some version checking / incrementation for my WordPress flow, and I need built assets and such before the actual release.

You can add the [@semantic-release/exec](/semantic-release/exec) to your release flow.
Then in the `prepare` step you can call a bashscript, or several in order to do your builds. You can get your version number via the `${nextRelease.version}` variable.

For a basic workflow you can see the [.releaserc](/oblakstudio/serbian-addons-for-woocommerce/blob/master/.releaserc) for this project.
